datasource/opennms-direct/src/main/java/org/opennms/oce/datasource/opennms/jvm/DirectInventoryDatasource.java

```diff
@@ -28,58 +28,172 @@
 import org.opennms.oce.datasource.common.HandlerRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
 
 /**
- * TODO: We need to periodically refresh the inventory here - as there
- * is currently no way for getting callbacks when the inventory changes.
+ * A datasource that provides {@link InventoryObject inventory} via the integration API.
  */
-public class DirectInventoryDatasource implements InventoryDatasource {
+public class DirectInventoryDatasource implements InventoryDatasource, AlarmLifecycleListener {
+    private static final Logger LOG = LoggerFactory.getLogger(DirectInventoryDatasource.class);
 
+    /**
+     * The registry of handlers interested in receiving callbacks for inventory.
+     */
     private final HandlerRegistry<InventoryHandler> inventoryHandlers = new HandlerRegistry<>();
 
-    private List<InventoryObject> inventoryObjects = new LinkedList<>();
+    /**
+     * The set of inventory objects initially populated via {@link #init} and subsequently via callbacks
+     * {@link #handleAlarmSnapshot} and {@link #handleNewOrUpdatedAlarm}.
+     */
+    private Set<InventoryObject> inventoryObjects = new LinkedHashSet<>();
+
+    /**
+     * A lock that prevents callbacks from being processed before we have finished {@link #init}.
+     */
+    private final CountDownLatch initLock = new CountDownLatch(1);
 
+    /**
+     * A {@link ReadWriteLock} to synchronize the readers and writers of {@link #inventoryObjects}.
+     */
     private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
 
+    /**
+     * Used during {@link #init} to initialize {@link #inventoryObjects}.
+     */
     private final NodeDao nodeDao;
 
-    public DirectInventoryDatasource(NodeDao nodeDao) {
+    /**
+     * Used during {@link #init} to initialize {@link #inventoryObjects}.
+     */
+    private final AlarmDao alarmDao;
+
+    /**
+     * @param nodeDao  used to retrieve the current inventory
+     * @param alarmDao used to retrieve the current inventory
+     */
+    public DirectInventoryDatasource(NodeDao nodeDao, AlarmDao alarmDao) {
         this.nodeDao = Objects.requireNonNull(nodeDao);
+        this.alarmDao = Objects.requireNonNull(alarmDao);
     }
 
+    /**
+     * On init we will populate the inventory by retrieving all current inventory from the {@link NodeDao} and
+     * {@link AlarmDao}.
+     */
     public void init() {
-        /// Populate the map with the current set of inventory
+        nodeDao.getNodes().forEach(n -> inventoryObjects.addAll(Mappers.toInventory(n)));
+        alarmDao.getAlarms().forEach(a -> processAlarm(a, false));
+        initLock.countDown();
+    }
+
+    /**
+     * Wait for {@link #init the init} to finish to delay callbacks from being processed and clients from polling for
+     * inventory.
+     */
+    private void waitForInit() {
```

Comment:
We should also call this function, or move this logic into the `waitUntilReady` method.